env0/resource_cost_credentials.go

```diff
@@ -0,0 +1,208 @@
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func resourceCostCredentials() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceCostCredentialsCreate,
+		ReadContext:   resourceCostCredentialsRead,
+		DeleteContext: resourceCostCredentialsDelete,
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:        schema.TypeString,
+				Description: "name for the credentials",
+				Required:    true,
+				ForceNew:    true,
+			},
+			"arn": {
+				Type:          schema.TypeString,
+				Description:   "the aws role arn",
+				Optional:      true,
+				ForceNew:      true,
+				ConflictsWith: []string{"client_id", "client_secret", "tenant_id", "subscription_id", "table_id", "secret"},
+				ExactlyOneOf:  []string{"client_id", "table_id"},
+			},
+			"external_id": {
+				Type:          schema.TypeString,
+				Description:   "the aws role external id",
+				Optional:      true,
+				Sensitive:     true,
+				ForceNew:      true,
+				RequiredWith:  []string{"arn"},
+				ConflictsWith: []string{"client_id", "client_secret", "tenant_id", "subscription_id", "table_id", "secret"},
+			},
+			"client_id": {
+				Type:          schema.TypeString,
+				Description:   "the azure client id",
+				Optional:      true,
+				ForceNew:      true,
+				ConflictsWith: []string{"external_id", "arn", "table_id", "secret"},
+				ExactlyOneOf:  []string{"arn", "table_id"},
+			},
+			"client_secret": {
+				Type:          schema.TypeString,
+				Description:   "azure client secret",
+				Optional:      true,
+				Sensitive:     true,
+				ForceNew:      true,
+				RequiredWith:  []string{"client_id"},
+				ConflictsWith: []string{"external_id", "arn", "table_id", "secret"},
+			},
+			"tenant_id": {
+				Type:          schema.TypeString,
+				Description:   "azure tenant id",
+				Optional:      true,
+				ForceNew:      true,
+				RequiredWith:  []string{"client_id"},
+				ConflictsWith: []string{"external_id", "arn", "table_id", "secret"},
+			},
+			"subscription_id": {
+				Type:          schema.TypeString,
+				Description:   "azure subscription id",
+				Optional:      true,
+				ForceNew:      true,
+				RequiredWith:  []string{"client_id"},
+				ConflictsWith: []string{"external_id", "arn", "table_id", "secret"},
+			},
+			"table_id": {
+				Type:          schema.TypeString,
+				Description:   "the table id of this credentials ",
+				Optional:      true,
+				ForceNew:      true,
+				ConflictsWith: []string{"client_id", "client_secret", "tenant_id", "subscription_id", "arn", "external_id"},
+				ExactlyOneOf:  []string{"arn", "client_id"},
+			},
+			"secret": {
+				Type:          schema.TypeString,
+				Description:   "the secret of this credentials",
+				Optional:      true,
+				Sensitive:     true,
+				ForceNew:      true,
+				RequiredWith:  []string{"table_id"},
+				ConflictsWith: []string{"client_id", "client_secret", "tenant_id", "subscription_id", "arn", "external_id"},
+			},
```

Comment:
Saw those generated nice on docs, just how did you do that?
I mean `google_cost_credentials` has only stuff that is related to GCP, my question is did you alter the docs files or did you manage to make it auto-generated like that?

Reply:
In order to have all the cloud providers in the same schema <b>and</b> and have the ability to set some of the arguments as  required (mutually exclusive), please consider having the following structure:

```
func resourceCostCredentials(cloudProvider string) *schema.Resource {
// This is an AWS only schema, same can be for GCP and Azure 
	awsSchema := map[string]*schema.Schema{
		"arn": {
			Type:          schema.TypeString,
			Description:   "the aws role arn",
			Required:      true,  // <------------------ Note the required attribute
			ForceNew:      true,
			ConflictsWith: []string{"client_id", "client_secret", "tenant_id", "subscription_id", "table_id", "secret"},
			ExactlyOneOf:  []string{"client_id", "table_id"},
		},
		"external_id": {
			Type:          schema.TypeString,
			Description:   "the aws role external id",
			Required:      true,
			Sensitive:     true,
			ForceNew:      true,
			RequiredWith:  []string{"arn"},
			ConflictsWith: []string{"client_id", "client_secret", "tenant_id", "subscription_id", "table_id", "secret"},
		},
	}
// This is the common part that will be added to every provider schema and in the end we'll have all the schemas in this schemas. 
	resultsSchema := map[string]*schema.Schema{
		"name": {
			Type:        schema.TypeString,
			Description: "name for the credentials",
			Required:    true,
			ForceNew:    true,
		},
	}
	
	if cloudProvider == "aws" {
		for index, element := range awsSchema {
			resultsSchema[index] = element
		}
	}

	return &schema.Resource{
		CreateContext: resourceCostCredentialsCreate,
		ReadContext:   resourceCostCredentialsRead,
		DeleteContext: resourceCostCredentialsDelete,
		Schema:        resultsSchema,
	}
```

Then in provider.go, we need to provide the cloud provider's name to the resource function.
E.g:
```
"env0_aws_cost_credentials":                 resourceCostCredentials("aws"),
"env0_azure_cost_credentials":               resourceCostCredentials("azure"),
"env0_google_cost_credentials":              resourceCostCredentials("gcp"),
```

Reply:
@sabasabo Its look more complex to me, the current code deal with all required arguments and the conflicts one. WDYT

Reply:
but the current code generates the docs wrong and then we need to manually alter them
which is problematic

Reply:
@samuel-br There are two issues we need to address:
- Having a single schema for 3 different cloud providers - you solved it 💪
- Enforcing the `Required` attribute in arguments that are essential for a cloud provider - the arguments in the code above are all `Optional` 😔

The second issue also affects the docs (I.e: `Optional` arguments instead of `Required` args).

Reply:
> Saw those generated nice on docs, just how did you do that? I mean `google_cost_credentials` has only stuff that is related to GCP, my question is did you alter the docs files or did you manage to make it auto-generated like that?
  
I didn't touch those files so I believe it generated automatically like that

Reply:
> * Enforcing the `Required` attribute in arguments that are essential for a cloud provider - the arguments in the code above are all `Optional` 😔

Yes they are, but the logic of `ConflictsWith` and `RequiredWith` creates the same results as it was `Required`, if required value wont set TF will address it as it was address to `Required` and throw error, you can find it in the UT for this resource, I am speaking from logic/technic perspective not from customer view(if you mean that it can confused the costumer its another problem). for the problem with the files @eranelbaz point on, I am little bit confuse here cause I don't understand how those files generated and I didn't change them, if the currents files are problematic LMK and I try to solve it, if it is @sabasabo you think what you suggested will solve it?

Reply:
> > Saw those generated nice on docs, just how did you do that? I mean `google_cost_credentials` has only stuff that is related to GCP, my question is did you alter the docs files or did you manage to make it auto-generated like that?
> 
> I didn't touch those files so I believe it generated automatically like that

I generated them locally and it changed the content to contain all the attributes for all of the providers